### ISSUE TITLE
Unexpected overload signature rejection since TypeScript 4.8.x

### ISSUE THREAD

**Issue by TimWolla** [microsoft/TypeScript#50754]
# Bug Report

### 🔎 Search Terms

- This overload signature is not compatible with its implementation signature.
- HTMLElementTagNameMap

### 🕗 Version & Regression Information

- This changed between versions 4.7.4 and 4.8.2

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.2&ssl=16&ssc=55&pln=16&pc=54#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVzQC0wBsATAIQE8AVAQwHMAeAaTlBmBRIGc4BVMMMChoaXYAwDWwChARwAElQCyAGQCiRYAFsOMWnQByNHUppgAfOYAUAKDitNOlDABccDdt0AaO3Bj0jHTcmHwBKN0VVDyc9AONgUzAAbRUIAHchETFmcwBdOAAfOBQkIiIAbhtQSFhEVAxsXAJicmp6K2BHXTdo7z84oLguGChMFDpw9y7nQuLSiqrwaHhkdCwcPEJSSn0O6dcpz2cvfsN4t2HR8cnemaKSsrgAb18EaDgrTXhMOABeOAADOU4D8GA4jjAAHTNUi3GAAYQgqBgwMwAGo0aFnr57Jg5HsIdCtiQoBwkphcpCUBASMBAsA-r9-v4zjosS97Jy4KSYEgoLhOoSYSSyRTKlyAL6+KW+Hl83APBYytA4YZwGjofDvf7CnbtEgQNBIGKQg1GmJwk4AcgAglbQpU8R8NQRoOzfCqUFwIJpIUQIHQrC6tVBIfhSQgHTYJeUgA)

### 💻 Code

```ts repro
export function childByTag<K extends Uppercase<keyof HTMLElementTagNameMap>>(
  element: Element,
  tagName: K,
): HTMLElementTagNameMap[Lowercase<K>] | null; // this overload is rejected
export function childByTag(element: Element, tagName: string): Element | null;
export function childByTag(element: Element, tagName: string): Element | null {
  for (let i = 0; i < element.childElementCount; i++) {
    if (element.children[i].nodeName === tagName) {
      return element.children[i];
    }
  }

  return null;
}

const anchor = childByTag(document.documentElement, 'A');
if (anchor) {
  console.log(anchor.href); // Would be rejected without the first overload.
};
```

### 🙁 Actual behavior

TypeScript rejected the first function overload, because "This overload signature is not compatible with its implementation signature".

### 🙂 Expected behavior

I didn't expect the code to be rejected, because:

- To my understanding all the values in `HTMLElementTagNameMap` are subtypes of `Element`, thus being a covariant return type.
- All the keys in `HTMLElementTagNameMap` are strings, thus matching a strict subset of all possible strings that may be passed to `tagName`. The remaining strings are matched by the second overload.

------

If the rejection is valid and my analysis is wrong, then I would have expected the error message to be more helpful in pointing out the problem.

**Comment by weswigham** [comment]
Right, so the change linked by the bot is essentially handling generic string mapping types more correctly, rather than just dropping them entirely. In most ways, it's right - contrary to what you may think, `HTMLElementTagNameMap[Lowercase<Uppercase<keyof HTMLElementTagNameMap>>]` is _not_ the same as `HTMLElementTagNameMap[keyof HTMLElementTagNameMap]` - for example, if the map contained a key where a member doesn't roundtrip through upper and lower-casing (like the İ or ß characters), you'd get a different key out than you put in. Given that, we don't "see" a `keyof HTMLElementTagNameMap` in that indexing position anymore, instead we see an opaque string-like type. So we calculate the constraint as equivalent to `HTMLElementTagNameMap[string]` (rather than the `HTMLElementTagNameMap["div" | "p" | "whatever"]` it was before), which, contrary to what you may think, does not evaluate, since `HTMLElementTagNameMap` has no index signatures and our constraint logic, unlike our comparison logic, requires index signatures exist to match open-ended types like `string`. Now, I've put up a PR (#50777) with a change to constraint calculation so `HTMLElementTagNameMap[string]` will resolve to the types of all the string-compatible keys - it doesn't seem to break much in our test suite and fixes this, so may be a good change.

**Comment by TimWolla** [comment]
@weswigham Thank you for writing up the detailed explanation of what is happening here. I understand some of those words :smiley: 

I don't want to take up more of your valuable time, but if you can answer: Based on what I understand, I can't rewrite the signature to be "more correct" / to work with 4.8.x, because of some limitation that is lifted by your PR?

**Comment by weswigham** [comment]
You can't rewrite it to be more correct, no, since I'd say it's already correct. But you can hack the typesystem to get the desired result with a workaround like
```ts
type Cast<A, B> = A extends B ? A : A & B;

export function childByTag<K extends Uppercase<keyof HTMLElementTagNameMap>>(
  element: Element,
  tagName: K,
): HTMLElementTagNameMap[Cast<Lowercase<K>, keyof HTMLElementTagNameMap>] | null;
```

**Comment by mosesoak** [comment]
I have a similar problem in this version. Here's an example signature with overloads as I originally wrote it. Note that there is a 2nd parameter, `whenKnown` which would either be not passed or be passed as `true` only.

```typescript
export function useMedia(mediaQuery: string, whenKnown: true): undefined | boolean;
export function useMedia(mediaQuery: string): boolean;
export function useMedia(mediaQuery: string, whenKnown?: true) { ... }
```

```typescript
export function useIsMobile(): boolean;
export function useIsMobile(whenKnown: true): undefined | boolean;
export function useIsMobile(whenKnown?: true) {
  return useMedia(mobleMqStr, whenKnown);
/*------------------------------^^^ 
Error: The call would have succeeded against this implementation, but implementation signatures of overloads are not externally visible.*/
}
```

It seems that because the inputs are variable / overloaded, it's finding errors with one or more of the callee overloads. Of course this works but hopefully it's not the only way to work around this error:

```typescript
  if (whenKnown) {
    return useMedia(mobileMqStr, true);
  }
  return useMedia(mobileMqStr);
```
So in the latest version, what's the correct way to do this?
